router/connection.go

```diff
@@ -472,46 +467,52 @@ func (conn *LocalConnection) receiveTCP(decoder *gob.Decoder, usingPassword bool
-			// payload is a subset of the receiver's topology, no
-			// further action is taken. Otherwise, the receiver sends
-			// out to all its connections an "improved" update:
-			//  - elements which the original payload added to the
-			//    receiver are included
-			//  - elements which the original payload updated in the
-			//    receiver are included
-			//  - elements which are equal between the receiver and
-			//    the payload are not included
-			//  - elements where the payload was older than the
-			//    receiver's version are updated
-			conn.SendTCP(Concat(ProtocolUpdateByte, conn.Router.Peers.EncodeAllPeers()))
-		} else if msg[0] == ProtocolUpdate {
-			newUpdate, err := conn.Router.Peers.ApplyUpdate(msg[1:])
-			if _, ok := err.(UnknownPeersError); err != nil && ok {
-				// That update contained a peer we didn't know about;
-				// request full update
-				conn.SendTCP(ProtocolFetchAllByte)
-				continue
+		} else if msg[0] == ProtocolPMTUVerified {
+			conn.verifyPMTU <- int(binary.BigEndian.Uint16(msg[1:]))
+		} else if msg[0] == ProtocolGossipUnicast {
+			origMsg := msg
+			channelHash, msg := decodeGossipChannel(msg[1:])
+			srcName, _, msg := decodePeerName(msg)
+			destName, _, msg := decodePeerName(msg)
+			if conn.local.Name == destName {
+				channel, found := conn.Router.GossipChannels[channelHash]
+				if !found {
+					conn.log("received unknown gossip channel:\n", channelHash)
+				} else {
+					channel.gossiper.OnGossipUnicast(srcName, msg)
+				}
+			} else {
+				conn.Router.Ourself.RelayGossipTo(destName, origMsg)
 			}
-			if conn.CheckFatal(err) != nil {
-				return
+		} else if msg[0] == ProtocolGossipBroadcast {
+			// intended for state from sending peer only
+			// done when there is a change that everyone should hear about quickly
+			// relayed using broadcast topology.
+			origMsg := msg
+			channelHash, msg := decodeGossipChannel(msg[1:])
+			srcName, _, msg := decodePeerName(msg)
+			channel, found := conn.Router.GossipChannels[channelHash]
+			if !found {
+				conn.log("received unknown gossip channel:\n", channelHash)
+			} else {
+				channel.gossiper.OnGossipBroadcast(msg)
 			}
-			if len(newUpdate) != 0 {
-				conn.Router.ConnectionMaker.Refresh()
-				conn.Router.Routes.Recalculate()
-				conn.Router.Ourself.BroadcastTCP(Concat(ProtocolUpdateByte, newUpdate))
+			conn.Router.Ourself.RelayGossipBroadcast(srcName, origMsg)
+		} else if msg[0] == ProtocolGossip {
+			// contains state for everyone that sending peer knows
+			// peers that receive it should examine the info, and if any of it is newer then
+			// pass it on to their peers
+			channelHash, msg := decodeGossipChannel(msg[1:])
+			channel, found := conn.Router.GossipChannels[channelHash]
+			if !found {
+				conn.log("received unknown gossip channel:\n", channelHash)
+			} else {
+				_, _, msg := decodePeerName(msg)
+				newBuf := channel.gossiper.OnGossip(msg)
+				if newBuf != nil {
+					channel.GossipMsg(newBuf)
+				}
```

Comment:
The above three gossip-related branches are crying out for a a helper invoked as `conn.handleGossip(msg[1:], func (channel, msg) {...})` that does the channel lookup, moans if it is not found, and otherwise invokes the function.